### ISSUE TITLE
scalar/integer miscompile from global isel for AArch64

### ISSUE THREAD

**Issue by regehr** [llvm/llvm-project#90245]
here's a function:
```llvm
define i32 @f(i32 %0, i32 %1) {
  %3 = sub i32 %1, %0
  %4 = sub nsw i32 %1, %0
  %5 = select i1 false, i32 %4, i32 %3
  %6 = icmp sgt i32 %5, 0
  %7 = select i1 %6, i32 0, i32 1
  ret i32 %7
}
```
using this driver:
```c
#include <stdio.h>

unsigned f(unsigned, unsigned);

int main(void) {
  printf("%x\n", f(0xe, 0x80000001));
}
```
it should print 0, and this is what happens with the sdag backend, but using the global isel backend it prints 1.

for the given inputs, `%4` is poison. the poison is blocked by the subsequent select but it seems like this is confusing the backend somehow.

cc @hatsunespica

**Comment by v01dXYZ** [comment]
Hi,

There's something strange I don't understand.

```
# *** IR Dump After IRTranslator (irtranslator) ***:
# Machine code for function f: IsSSA, TracksLiveness
Function Live Ins: $w0, $w1

bb.1 (%ir-block.2):
  liveins: $w0, $w1
  %0:_(s32) = COPY $w0
  %1:_(s32) = COPY $w1
  %4:_(s1) = G_CONSTANT i1 false
  %6:_(s32) = G_CONSTANT i32 0
  %9:_(s32) = G_CONSTANT i32 1
  %2:_(s32) = G_SUB %1:_, %0:_
  %3:_(s32) = nsw G_SUB %1:_, %0:_
  %5:_(s32) = G_SELECT %4:_(s1), %3:_, %2:_
  %7:_(s1) = G_ICMP intpred(sgt), %5:_(s32), %6:_
  %8:_(s32) = G_SELECT %7:_(s1), %6:_, %9:_
  $w0 = COPY %8:_(s32)
  RET_ReallyLR implicit $w0

# End machine code for function f.
```

The combiner does something very fishy:

```
Try combining %5:_(s32) = G_SELECT %4:_(s1), %3:_, %2:_
10: GIM_SwitchOpcode(MIs[0], [19, 274), Default=6760, JumpTable...) // Got=142
3921: Begin try-block
3928: GIM_CheckSimplePredicate(Predicate=51)
3932: GIM_CheckCxxPredicate(MIs[0], Predicate=23)
3935: GIR_CustomAction(FnID=15)
Erasing: %5:_(s32) = G_SELECT %4:_(s1), %3:_, %2:_

Changing: %7:_(s1) = G_ICMP intpred(sgt), %5:_(s32), %6:_

Changed: %7:_(s1) = G_ICMP intpred(sgt), %3:_(s32), %6:_

3936: GIR_Done
```

**Comment by v01dXYZ** [comment]
I think I've got the reason. It seems the two subs are seen as equivalent (ie `~= select false, A, A`) which makes the instruction match the `select_same_val` pattern in `Combine.td`. cf `CodeGen/GlobalISel/CombinerHelper.cpp matchSelectSameVal` and `matchEqualDefs`. Beneath it uses `produceSameValue(...) -> MachineInstr::isIdentical`. The code is at:

https://github.com/llvm/llvm-project/blob/b2c9f7d3188e41163574a83a835437955cf4b80f/llvm/lib/CodeGen/MachineInstr.cpp#L642

I don't know if it is a fault as marking one as `nsw` could also imply the other one as `nsw`. BUT swapping the two operands produce different outputs, so in this sense the two should not be seen as equivalent.

**Comment by tschuett** [comment]
From a brief scan of `isDenticalTo`, it looks as if it ignores flags.

**Comment by dtcxzyw** [comment]
Any progress? Looks like this issue is still standing: https://godbolt.org/z/7s3hYbh4z
cc @osa1 @davemgreen

**Comment by davemgreen** [comment]
It looks like CombinerHelper::matchEqualDefs or isIdenticalTo is not considering flags, so the select folds two almost-identical subs to the wrong instruction.

**Comment by osa1** [comment]
`CombinerHelper::matchEqualDefs` doesn't check the flags but that may be intentional (hard to tell as it's not documented), and also making it check the flags doesn't give us the optimized version generated by sdag when the `G_SELECT` condition is not constant.

Here are the tests: (the difference between the two is the `G_SELECT` operand order)

```
---
name:            select_dynamic_1
tracksRegLiveness: true
body:             |
  bb.1:
    liveins: $w0, $w1, $w2

    %0:_(i32) = COPY $w0
    %1:_(i32) = COPY $w1
    %2:_(i32) = COPY $w2
    %3:_(i1)  = G_TRUNC %2(s32)
    %4:_(i32) = G_SUB %1:_, %0:_
    %5:_(i32) = nsw G_SUB %1:_, %0:_
    %6:_(i32) = G_SELECT %3:_(i1), %5:_, %4:_
    $w0 = COPY %6:_(i32)
    RET_ReallyLR implicit $w0
...

---
name:            select_dynamic_2
tracksRegLiveness: true
body:             |
  bb.1:
    liveins: $w0, $w1, $w2

    %0:_(i32) = COPY $w0
    %1:_(i32) = COPY $w1
    %2:_(i32) = COPY $w2
    %3:_(i1)  = G_TRUNC %2(s32)
    %4:_(i32) = G_SUB %1:_, %0:_
    %5:_(i32) = nsw G_SUB %1:_, %0:_
    %6:_(i32) = G_SELECT %3:_(i1), %4:_, %5:_
    $w0 = COPY %6:_(i32)
    RET_ReallyLR implicit $w0
...
```

In both cases we want to see one `G_SUB` without the `nuw` flag. Just making `matchEqualDefs` check the flags won't give us that. (it'll leave both `G_SUB`s and `G_SLECT`)

The original repro above is easy to fix as it's selecting based on a constant, so `select_constant_cmp` will take care of it if we just make `matchEqualDefs` check the flags, but that won't fix the more general case or non-constant `G_SELECT` condition with same operands differing just in flags.

LLVM IR for the MIR tests above:

```
define i32 @f(i32 %a, i32 %b, i1 %c) {
  %sub = sub i32 %b, %a
  %sub.nsw = sub nsw i32 %b, %a
  %sel = select i1 %c, i32 %sub.nsw, i32 %sub
  %cmp = icmp sgt i32 %sel, 0
  %res = select i1 %cmp, i32 0, i32 1
  ret i32 %res
}
```

I looked into how we might go about replacing two instructions differing only in flags with one with the LUB of the flags (i.e. take the conservative ones of differing flags), but it looks like we don't have this API readily available for machine instructions, so that needs to be implemented first. I'm happy to do it myself if a resident gisel expert can confirm that the approach makes sense. (I'm kinda new so I'm not sure myself)

**Comment by lonemeow** [comment]
A side observation: the flag-blindness is missing in the documentation at every level of the chain this goes through (`CombinerHelper::matchEqualDefs()` -> `TargetInstrInfo::produceSameValue()` -> `MachineInstr::isIdenticalTo()`):

  ```c++
    /// Return true if this instruction is identical to \p Other.
    /// Two instructions are identical if they have the same opcode and all their
    /// operands are identical (with respect to MachineOperand::isIdenticalTo()).
    /// Note that this means liveness related flags (dead, undef, kill) do not
    /// affect the notion of identical.
    LLVM_ABI bool isIdenticalTo(const MachineInstr &Other,
                                MICheckType Check = CheckDefs) const;
  ```
  ```c++
    /// Return true if two machine instructions would produce identical values.
    /// By default, this is only true when the two instructions
    /// are deemed identical except for defs. If this function is called when the
    /// IR is still in SSA form, the caller can pass the MachineRegisterInfo for
    /// aggressive checks.
    virtual bool produceSameValue(const MachineInstr &MI0,
                                  const MachineInstr &MI1,
                                  const MachineRegisterInfo *MRI = nullptr) const;
  ```

These appear to document explicitly what they don't compare, but the lists are no longer complete: MI flags are also not compared, and those can affect the value produced.

It's worth noting that the IR side makes the flag comparison explicit all the way in the function name in `Instruction::isIdenticalTo()` vs `Instruction::isIdenticalToWhenDefined()`.
